File: COS382/Comapring_Languages/csharp/prime.cs
//!/usr/bin/mono

using System;
using System.Collections.Generic;
using System.Linq;

namespace PrimePartition{
  class Prime{
    static void Main(string[] args){
      List<int> prime = new List<int>();
      Console.WriteLine(" Prime Partitions Program");
      Console.WriteLine("Enter a number: ");
      int a = Convert.ToInt32(Console.ReadLine());
      for (int i = 0; i <= a; i++){
        if (is_prime(i) == true){
          prime.Add(i);
        }
      }
      prime.ToArray();
      Console.WriteLine("Primes up to "+a+": "+string.Join(" ",prime));
      List<int> primes = new List<int>();
      prime_part_constructive(a, 1, primes);
    }
    static bool is_prime(int n){
      if (n <= 1) return false;
      for (int i = 2; i <= Math.Sqrt(n); i++){
        if (n % i == 0) return false;
      }
      return true;
    }
    static void prime_part_constructive(int n, int k, List<int> lst){
      if (n == 0){
        //Console.WriteLine(n);
        lst.ToArray();
        Console.WriteLine(string.Join("+", lst));
        //Console.WriteLIne(lst);
        //lst.ForEach(Console.WriteLine);
        //Console.WriteLine(lst.ToArray());
      }
      if (n <= k){
        return;
      }
      for (int i = (k+1); i <= (n+1); i++){
        if (is_prime(i) == false){
          continue;
        }
        lst.Add(i);
        prime_part_constructive(n-i, i, lst);
        lst.Remove(i);
      }
    }
  }
}

File: COS382/Comapring_Languages/ruby%20/prime.rb
#!/usr/bin/ruby

def is_prime(n)
	if n <= 1
		return false
	end
	for i in 2...n
		if n % i == 0
			return false
		end
	end
	return true
end

def prime_part_constructive(n, k, lst)
	if n == 0
		puts lst.join("+")
	end
	if n <= k
		return
	end
	for i in (k+1)..(n+1)
		if not is_prime(i)
			next
		end
		prime_part_constructive(n-i, i, lst << i)
		lst.pop
	end
end

puts "Prime Partitions Program"
print "Enter a number: "
a = gets.to_i

primes = []
for i in 0..a
	if is_prime(i)
		primes << i
	end
end

puts "Primes up to #{a}: #{primes.inspect}"

prime_part_constructive(a, 1, [])
puts ""
